Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -564,10 +564,16 @@ public struct ObjectId: BSONValue, Equatable, CustomStringConvertible, Codable {
         }
     }
 
+    /// Returns the provided string as a `bson_oid_t`.
+    internal static func encode(_ str: String) -> bson_oid_t {
```

Comment:
I'm not sure it makes sense to introduce a new static function instead of sticking with 2 LOC in the two places this would be called:

```
var oid = bson_oid_t()
bson_oid_init_from_string(&oid, self.oid)
```

Part of my objection is also with the name itself implying that it's an overload, until you realize it's very different than the `encode()` protocol method.

Reply:
This is something I spoke to Kaitlin about in [#swift](https://mongodb.slack.com/archives/C89CNDNF6/p1541095355137600).

I suppose I'll tag Kaitlin here to see if we want to actually not do this after all @kmahar.

Reply:
I think we should keep this abstraction to be consistent with `Decimal128`, which is the only other type where we have to do some kind of conversion b.w our storage format and libbson's. (in that case it's 4 lines that are called 3x, so we should definitely keep that.)

that said, I would support renaming these methods, to something more descriptive that clarifies what we're doing. `convertToLibbsonType()` or something? idk.

Reply:
I'm not sure about Decimal128's `encode()` method, but given that this is always called with an ObjectID instance in mind, I'd propose we remove `static` and make this an actual internal method, such as `toInternalType()` with no arguments.

I would suggest a computed property, but Decimal128 initialization does throw. And we can't assume its internal data is valid because there is an second `init()` method that doesn't validate and trusts the user provided something valid.

Reply:
I left the office before I could really respond to Jeremy's points above, but I gave him the very solid thumbs up. I think that's a good way to meet in the middle here, and the proposed changes of making the method internal sounds good.

I am not completely sure about the removal of the arguments however, because it seems like this method is at some point being called in a `static` manner, for example, in the [`Decimal128`'s `init?()`](https://github.com/mongodb/mongo-swift-driver/blob/master/Sources/MongoSwift/BSON/BSONValue.swift#L318).

However, from what I can tell that is the _only_ time we use the method in a `static` fashion, both for `Decimal128` and `ObjectId`.

Reply:
ya, the Decimal128 one is static so it can be used from within an init to validate data before actually going ahead with creating the object.
all of this making me look at libbson's [bson_oid_t](http://mongoc.org/libbson/current/bson_oid_t.html) and I'm a little confused now - should we be calling `bson_oid_is_valid` ever? does `bson_oid_init_from_string` ever fail / how can we tell?

Reply:
@kmahar had the same kind of questions about `bson_oid_init_from_string()` failing, but did not notice a `bson_oid_is_valid()`.

It seems like if such a function for checking validity exists, we are supposed to using it. I'll see if perhaps Kevin has an idea on this and get back to this thread.

Reply:
Did a rename of `encode()` to `toLibBSONType()`.

Reply:
@jmikola Answered this in #mongoc:
> PHP does use is_valid: https://github.com/mongodb/mongo-php-driver/blob/master/src/BSON/ObjectId.c#L49

And:
> checks the string buffer length and validates its characters: https://github.com/mongodb/mongo-c-driver/blob/master/src/libbson/src/bson/bson-oid.c#L249

Seems like we'd do well to follow PHP here and also validate the `oid` strings. I'll see if I can do a search for our uses of `bson_init_from_string()` and make sure it follows a validity check.

Reply:
I added a validation check to our call to `bson_init_from_string()`. In order to stay consistent with something like `Decimal128`, I've also added a failable initializer to `ObjectId`.

Reply:
lol oops, forgot we still had this discussion to resolve before approving, new changes also LGTM  tho

Reply:
> ya, the Decimal128 one is static so it can be used from within an init to validate data before actually going ahead with creating the object.

Missed that. Makes sense for Decimal128 to be static then. I'm ambivalent about ObectId.